doxygen/man/man3/TransactionFilterProxy.3
.TH "TransactionFilterProxy" 3 "Wed Feb 10 2016" "Version 1.0.0.0" "darksilk" \" -*- nroff -*-
.ad l
.nh
.SH NAME
TransactionFilterProxy \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <transactionfilterproxy\&.h>\fP
.PP
Inherits QSortFilterProxyModel\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBTransactionFilterProxy\fP (QObject *parent=0)"
.br
.ti -1c
.RI "void \fBsetDateRange\fP (const QDateTime &from, const QDateTime &to)"
.br
.ti -1c
.RI "void \fBsetAddressPrefix\fP (const QString &addrPrefix)"
.br
.ti -1c
.RI "void \fBsetTypeFilter\fP (quint32 modes)"
.br
.ti -1c
.RI "void \fBsetMinAmount\fP (const \fBCAmount\fP &minimum)"
.br
.ti -1c
.RI "void \fBsetLimit\fP (int limit)"
.br
.ti -1c
.RI "void \fBsetShowInactive\fP (bool showInactive)"
.br
.ti -1c
.RI "int \fBrowCount\fP (const QModelIndex &parent=QModelIndex()) const "
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static quint32 \fBTYPE\fP (int type)"
.br
.in -1c
.SS "Static Public Attributes"

.in +1c
.ti -1c
.RI "static const QDateTime \fBMIN_DATE\fP = QDateTime::fromTime_t(0)"
.br
.ti -1c
.RI "static const QDateTime \fBMAX_DATE\fP = QDateTime::fromTime_t(0xFFFFFFFF)"
.br
.ti -1c
.RI "static const quint32 \fBALL_TYPES\fP = 0xFFFFFFFF"
.br
.in -1c
.SS "Protected Member Functions"

.in +1c
.ti -1c
.RI "bool \fBfilterAcceptsRow\fP (int source_row, const QModelIndex &source_parent) const "
.br
.in -1c
.SH "Detailed Description"
.PP 
Filter the transaction list according to pre-specified rules\&. 
.PP
Definition at line 10 of file transactionfilterproxy\&.h\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "TransactionFilterProxy::TransactionFilterProxy (QObject * parent = \fC0\fP)\fC [explicit]\fP"

.PP
Definition at line 13 of file transactionfilterproxy\&.cpp\&.
.SH "Member Function Documentation"
.PP 
.SS "bool TransactionFilterProxy::filterAcceptsRow (int source_row, const QModelIndex & source_parent) const\fC [protected]\fP"

.PP
Definition at line 25 of file transactionfilterproxy\&.cpp\&.
.SS "int TransactionFilterProxy::rowCount (const QModelIndex & parent = \fCQModelIndex()\fP) const"

.PP
Definition at line 86 of file transactionfilterproxy\&.cpp\&.
.SS "void TransactionFilterProxy::setAddressPrefix (const QString & addrPrefix)"

.PP
Definition at line 57 of file transactionfilterproxy\&.cpp\&.
.SS "void TransactionFilterProxy::setDateRange (const QDateTime & from, const QDateTime & to)"

.PP
Definition at line 50 of file transactionfilterproxy\&.cpp\&.
.SS "void TransactionFilterProxy::setLimit (int limit)"
Set maximum number of rows returned, -1 if unlimited\&. 
.PP
Definition at line 75 of file transactionfilterproxy\&.cpp\&.
.SS "void TransactionFilterProxy::setMinAmount (const \fBCAmount\fP & minimum)"

.PP
Definition at line 69 of file transactionfilterproxy\&.cpp\&.
.SS "void TransactionFilterProxy::setShowInactive (bool showInactive)"
Set whether to show conflicted transactions\&. 
.PP
Definition at line 80 of file transactionfilterproxy\&.cpp\&.
.SS "void TransactionFilterProxy::setTypeFilter (quint32 modes)"

.PP
\fBNote:\fP
.RS 4
Type filter takes a bit field created with \fBTYPE()\fP or ALL_TYPES 
.RE
.PP

.PP
Definition at line 63 of file transactionfilterproxy\&.cpp\&.
.SS "static quint32 TransactionFilterProxy::TYPE (int type)\fC [inline]\fP, \fC [static]\fP"

.PP
Definition at line 24 of file transactionfilterproxy\&.h\&.
.SH "Member Data Documentation"
.PP 
.SS "const quint32 TransactionFilterProxy::ALL_TYPES = 0xFFFFFFFF\fC [static]\fP"
Type filter bit field (all types) 
.PP
Definition at line 22 of file transactionfilterproxy\&.h\&.
.SS "const QDateTime TransactionFilterProxy::MAX_DATE = QDateTime::fromTime_t(0xFFFFFFFF)\fC [static]\fP"
Last date that can be represented (far in the future) 
.PP
Definition at line 20 of file transactionfilterproxy\&.h\&.
.SS "const QDateTime TransactionFilterProxy::MIN_DATE = QDateTime::fromTime_t(0)\fC [static]\fP"
Earliest date that can be represented (far in the past) 
.PP
Definition at line 18 of file transactionfilterproxy\&.h\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for darksilk from the source code\&.


CHANGELOG
Base58 Change, Port Changes and Minor Fixes
Addition of OBJ directory for Daemon builds
Version updated
ReadMe updated
Instamine code for Silk/Dark weaver swap/burn added to main.cpp
OpenSSL update
LevelDB 1.18 Update
Fix to guard against OpenSSL's new strict DER checks, fixed getbalance and other minor fixes
Spelling edits
Enabled locking of Stormnodes via the Stormnode configuration file
Unset Sandstorm checkbox if there were non-anonymized inputs selected in coincontrol, give warning
Removal of target space change and hardfork
Fixed time to switch "out of sync" messages
Fix twice denominatedBalance calculations
Solve performance problems by caching and fix probable security issue
RPC: Allow "sandstorm" only with ENABLE_WALLET
Disconnect Sandstorm timer on Overview page destruction
Fix spaces/monospace fonts in qt console
Mac compatibility so users can switch tabs using CMD+(NUM)
Added better instructions to get libsecp256k1 installed in case of library issues within ReadMe
Added secp256k1 libs to /deps in case for any reason the libs are not available from github anymore
Increased maximum block size to 50MB (50x Bitcoin Core)
Added safe shutdown for Windows
Fixed sync issues and CPU usage
Add event support for Tor
rpc: Prevent easy memory exhaustion attack 
Fixed issue where using setaccount with a foreign address causes the address to be added to your receiving addresses.
Fix priority calculation in CreateTransaction
darksilk.conf added to /src
GUI Styling changes
Statistics Page Added and Icons Updated
Clickable Padlock to Unlock Wallet and Unlock for Staking Only
Fixed GUI bugs
Assertion Error Fix for MacOSX
Sandstorm GUI Fixes
Decentralised DarkSilkMarket added
Overview Transactions UI Fixes
Messaging, GUI Update, Null Pointer Fix, Various Fixes/Changes
Windows Build Fix
Addition of possibly needed dependency
Stormnode Collateral Increase
Marketplace RPC Calls | Cleanup
Amended Spork Max Value for Stormnode Collateral
Removal of unused Spork
Minimum Transaction fee changed to 0.0001DRKSLK
Add support for miniupnpc API version 14
DARKSILK_TIMEDATA_MAX_SAMPLES 200
Add time offset to getpeerinfo output
Don't check duplicate stake while importing
Ignore blocks received while importing
Further Market RPC JSON calls
marketbuyrequest RPC command
marketmybuys RPC call for market
MacOSX Fixes and GUI Fix
Market RPC Minor Fixes
darksilk-qt.pro cleanup
DarkSilk-qt.app renamed to DarkSilk.app for OSX
Replaced stake min age with min confirmation
Limited drop of difficulty per block
Made special marker non spendable
Added nonspendable encoding for block signing key
BIP66 compatibility
Replaced NOP7 with CHECKLOCKTIMEVERIFY
Enforce new script verification flags
Static pos reward
Demand canonical block sig from updated peers
Submitblock: Add submission of stake template
Reward and Min Conf Blocks Changed
Checkpointing Removal
Gitian MacDeploy Graphics Added
IsProtocolV1/IsProtocolV2 hard fork switches removed
FutureDriftV1/FutureDriftV2 reverted back to FutureDrift
CheckStakeKernelHashV1/CheckStakeKernalHashV2 reverted back to CheckStakeKernelHash
Unsigned Int GetTargetSpacing changed to Static Const Int POS_TARGET_SPACING
Fix boost::get usage with Boost 1.58 in rpcrawtransaction.cpp 
Static PoS Reward of 1 DRKSLK
Prevent redefinition compiler warning
Fix issue where getrawtransaction RPC call does not work
PoW additions to GetNextTargetRequired
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
LevelDB windows cross compile missing files
Scale up addrman
Gitian windows compile fixes
Correctly randomize change output position
Removal of Dead Code
Fix for building with Boost 1.58
LevelDB Cross compile fixes
DarkSilk-Qt: cleanup / optimise addressbookpage
Discourage fee sniping with nLockTime
Removal of dead code from main.h
Changed ProofOfStakeReward to use static const
Diff changes
Shadowchat Messaging Updates
DNSSeed Thread Fix
Stormnode payments start on block 82002
Restore hex to getrawtransaction vout scriptPubkey
static const rename | nActualSpacing code reform
Removal of Keepass until later date that this doesn't cause an issue with Windows builds
Updated dependencies, added Gitian OSX files
Remove unused GetProofOfStakeReward
Faster Retargetting for PoW
Fixed PoS target spacing
Fix payee selectoin on StormnodePayments::ProcessBlock
fix log output / more notCapableReason-s
HasCollateralInputs should check only for at least 1 collateral not 2
Track lastTimeChanged for mnodeman and ask for update only if UpdateNeeded
Added sstx support for free transactions 
Fix for src/qt/stormnodemanager.cpp
RPC Fixes
Fix DoAutomaticDenominating
Few Fixes for Processing Extra Messages
Reversioning and restriction change
Check and remove expired Stormnodes on Client start
Maps in snodeman
bool warning fix
SSEG Update
Show new number of SN's on add/remove in logs | Move output to debug level
sigTime
Temp Fix prior to major changes
SPORK Additions/Changes | SANDSTORM fee removal
prevent ganing sigTime too much on initial ssee
SandStorm/InstantX Functionality | Send with SandStorm on sendcoinsdialog.ui
Fix Filtering for active | add spaces to table-fy console view | add spaces to IP only in full mode
Merged Stormnode blinding
fix compiling for windows
fix console spam
Make stormnodes.dat verification more specific, prevent from overwriting corruped
files with unknown format, give more info in debug.log output
fixed shared key usage
fix even more console spam
rename stormnodes.dat->mncache.dat / fix strings
save nsqCount in mncache.dat
[Qt] add more NULL pointer checks in darksilkgui.cpp
close wallet regardless of "minimize to tray" settings
Added coincontrol sandstorm rounds column
Stormnode Proof of Service Added
Add structure for donating a percentage of ones earnings to another party via the protocol
Fixed snw message issue
fix pubkey/donation conditions
Fixed blinding submission process
increasing ss timeouts
fixed ssee sync
fixed Stormnode payments FindOldestNotInVec

src/qt/transactionfilterproxy.cpp
#include "transactionfilterproxy.h"
#include "transactiontablemodel.h"
#include "transactionrecord.h"

#include <QDateTime>
#include <cstdlib>

// Earliest date that can be represented (far in the past)
const QDateTime TransactionFilterProxy::MIN_DATE = QDateTime::fromTime_t(0);
// Last date that can be represented (far in the future)
const QDateTime TransactionFilterProxy::MAX_DATE = QDateTime::fromTime_t(0xFFFFFFFF);

TransactionFilterProxy::TransactionFilterProxy(QObject *parent) :
    QSortFilterProxyModel(parent),
    dateFrom(MIN_DATE),
    dateTo(MAX_DATE),
    addrPrefix(),
    typeFilter(ALL_TYPES),
    minAmount(0),
    limitRows(-1),
    showInactive(true)
{
}

bool TransactionFilterProxy::filterAcceptsRow(int sourceRow, const QModelIndex &sourceParent) const
{
    QModelIndex index = sourceModel()->index(sourceRow, 0, sourceParent);

    int type = index.data(TransactionTableModel::TypeRole).toInt();
    QDateTime datetime = index.data(TransactionTableModel::DateRole).toDateTime();
    QString address = index.data(TransactionTableModel::AddressRole).toString();
    QString label = index.data(TransactionTableModel::LabelRole).toString();
    qint64 amount = llabs(index.data(TransactionTableModel::AmountRole).toLongLong());
    int status = index.data(TransactionTableModel::StatusRole).toInt();

    if(!showInactive && (status == TransactionStatus::Conflicted || status == TransactionStatus::NotAccepted))
        return false;
    if(!(TYPE(type) & typeFilter))
        return false;
    if(datetime < dateFrom || datetime > dateTo)
        return false;
    if (!address.contains(addrPrefix, Qt::CaseInsensitive) && !label.contains(addrPrefix, Qt::CaseInsensitive))
        return false;
    if(amount < minAmount)
        return false;

    return true;
}

void TransactionFilterProxy::setDateRange(const QDateTime &from, const QDateTime &to)
{
    this->dateFrom = from;
    this->dateTo = to;
    invalidateFilter();
}

void TransactionFilterProxy::setAddressPrefix(const QString &addrPrefix)
{
    this->addrPrefix = addrPrefix;
    invalidateFilter();
}

void TransactionFilterProxy::setTypeFilter(quint32 modes)
{
    this->typeFilter = modes;
    invalidateFilter();
}

void TransactionFilterProxy::setMinAmount(const CAmount& minimum)
{
    this->minAmount = minimum;
    invalidateFilter();
}

void TransactionFilterProxy::setLimit(int limit)
{
    this->limitRows = limit;
}

void TransactionFilterProxy::setShowInactive(bool showInactive)
{
    this->showInactive = showInactive;
    invalidateFilter();
}

int TransactionFilterProxy::rowCount(const QModelIndex &parent) const
{
    if(limitRows != -1)
    {
        return std::min(QSortFilterProxyModel::rowCount(parent), limitRows);
    }
    else
    {
        return QSortFilterProxyModel::rowCount(parent);
    }
}
